apps/jupyter/extensions/notebook-runner/src/packages.ts

```diff
@@ -0,0 +1,208 @@
+import { JupyterFrontEnd } from "@jupyterlab/application";
+import { ISessionContext } from "@jupyterlab/apputils";
+import { INotebookTracker } from "@jupyterlab/notebook";
+import { IKernelConnection } from "@jupyterlab/services/lib/kernel/kernel";
+import { Contents, ContentsManager, KernelMessage } from "@jupyterlab/services";
+import { addKernelListeners, writeJsonToVirtualFilesystem } from "./utils";
+import { EmbeddedPythonCallbacks } from "./iframe-api";
+
+const autoInstallPackages =
+  (contentsManager: ContentsManager, notebookPath: string) =>
+  async (kernel: IKernelConnection): Promise<void> => {
+    kernel.iopubMessage.connect((_, msg) => {
```

Comment:
Why does this function need to be defined inline? It's rather long, it might be better off on its own.